### DISCUSSION TITLE
Isolated Run - Side Effects

### DISCUSSION THREAD

**Discussion by ghost** [h3rald/min#120] [Features: On Hold]
Does min supports isolated dequoting and applying or isolated symbols?

Because test schedule needs isolated running, ex.:

```
(<some file applications (side effects included!)>) assert ; -> this assert must test quotes isolatedly
```

**Comment by h3rald** [comment]
Uhm... yes: `apply` by definition evaluates a quotation in a separate stack. So yes probably I should use it for tests -- the test module is probably the oldest min code that I wrote, and I had to change it this week because of 0.31.0, but still it is not a great example of proper min code with the latest idioms and best practices.

**Reply by ghost** [reply]
> Uhm... yes: `apply` by definition evaluates a quotation in a separate stack. So yes probably I should use it for tests -- the test module is probably the oldest min code that I wrote, and I had to change it this week because of 0.31.0, but still it is not a great example of proper min code with the latest idioms and best practices.

Isolated run problem who i thought is not only for `test` module. Maybe you only want to get the type of quote after dequoting process:

```
( symbol isolated
  (quot :q ==> bool :r)
  (q -> type "string" == @r) ; -> Side-effects 🥵! I cannot do any file operations now.
) operator
```

**Reply by ghost** [reply]
Maybe we need more symbols that are not existing in min:
`operator-inputs`, `operator-outputs`, `operator-paramquot` ... These would make more actions on operator/typeclass/constructor signatures and also would make isolation of type-pred symbols.

**Reply by h3rald** [reply]
I understand what you are saying, but sometimes you want side effects... maybe another type of `function` operator that creates a symbol with no side effects by taking only input values from the stack and then evaluating its body in a separate stack?

You should more or less be already able to do that by wrapping most of the body of your operator in a quotation and then push `apply` to evaluate it in a separate stack. Did you try that?

**Reply by ghost** [reply]
Okay look at example below:

```
( typeclass dequo_bool ; Side Effects Included!
    (quot :q ==> bool :r)
    (q -> type "bool" == @r)
) ::
```

The above typeclass has side-effects. Because dequoting may make occurrence of side-effects. So there is no way to test type of dequoting result, i think.

**Reply by ghost** [reply]
> but sometimes you want side effects...

but sometimes you don't want side effects. One may only want snatch the type of quote not one want running all quote for example.

A qualifier should be added `no-side-effects` to min.

**Reply by ghost** [reply]
>  evaluate it in a separate stack

Evaluating it where in a separate stack evaluates file reading-writing too. More stacking can't be a solution, sorry. 😭

**Reply by h3rald** [reply]
OK, wait, you are talking *generally* about forbidding side effects like writing to file, or even printing to the screen Haskell-style?

Or maybe even declaring which side effects are allowed? This is going to be tricky, and I wouldn't want to make the language too heavy.

How would you go about implementing it in min though? Except for blacklisting/whitelisting some symbols maybe? It would require quite a lot of existing symbols though, to declare what side effects they cause.

**Reply by ghost** [reply]
> Okay look at example below:
> 
> ```
> ( typeclass dequo_bool ; Side Effects Included!
>     (quot :q ==> bool :r)
>     (q -> type "bool" == @r)
> ) ::
> ```
> 
> The above typeclass has side-effects. Because dequoting may make occurrence of side-effects. So there is no way to test type of dequoting result, i think.

I want to use typeclasses in my code as no-side-effected. I want to know: Do they produce side-effects?
